### DISCUSSION TITLE
Plugin for arduino

### DISCUSSION THREAD

**Discussion by seb5g** [PyMoDAQ/PyMoDAQ#179] [General]
This chat will be about putting together exemples of use of an arduino as an instrument plugin for PyMoDAQ. The discussion started in french by email. I will copy the exchanges below but please keep going in English .

**Comment by seb5g** [comment]
je vous propose de créer (Jérémie) un plugin: pymodaq_plugins_arduino

Dans lequel on mettra nos exemples d'application pymodaq avec Arduino par exemple ceux de Jérémie avec télémétrie... Ceux de Bernhard...

Avec des noms clairs

**Comment by seb5g** [comment]
Answer from Bernhard Lang:

créer un plugin Arduino générique : selon ce qu'on a lui codé dessus et soudé autour, il peut représenter à la fois une machine à café, un distributeur de bière, une distillerie de kirsch, la fabrication du Roquefort etcpp, si vous voyez ce que veux dire. Et le langage à lui parler pourra être tout entre le patagonien et le sibérien oriental.

Lors qu'on utilise une librairie comme telemetrix pour piloter l'Arduino, comme proposé par Jérémie pour le tutoriel, ce sera en fin un tutoriel comment parler à telemetrix par Pymodaq avec un Arduino derrière comme appareil pas cher. On devrait donc, correctement dit, l'appeler plutôt Pymodaq_plugins_telemetrix_arduino. Et en passant par la librairie meletetrix ce sera Pymodaq_plugins_meletetrix_arduino.

voilà ce qui concerne les noms clairs 🙂

Ce qui sera commun à tous les cas dont l'Arduino est utilisé de façon "bare metal", c'est la communication avec l'appareil par ligne serielle (très probablement virtuelle).

deux idées : créer un plugin qui, par son nom, déclare son caractère éducatif, pour mettre tout-ce là-dedans qui pourrait être utile pour apprendre étape par étape

inspiré par un projet d'un ancien collègue, un arduino ou esp ou raspberry ou n'importe quel bestiole semblable, l'essentiel est qu'on le programme soi même, annonce ces capabilités (comme move, viewer et attributes) lors de la connection par un simple protocole à définir. Et le plugin générique correspondant les annonce à PyMoDAQ. Le rêve de plug&play du bricoleur en électronique :-)

Bernhard

**Comment by seb5g** [comment]
Answer from Sebastien:

On tombe dans le même cas de figure que le plugin daqmx. On peut y mettre un peu tout ce qu'on y veut selon ce qui est branché dessus. Pour autant il contient un wrapper python qui simplifie les développements en python et qqs exemples d'utilisation. Mais quand j'utilise des fonctionnalités de daqmx dans mes manips, je crée en fait un plugin très custom (pour faire du café par exemple ;-) - avec probablement uniquement moi qui pourrait l'utiliser) par exemple j'en ai un qui s'appelle pymodaq_plugins_moke (magneto-optical kerr effect) qui utilise en sous main le wrapper daqmx pour acquérir, synchroniser illumination et caméra etc...Mais je ne le partage pas ni ne le publie... 

Donc selon moi un plugin arduino ne peut contenir que des exemples variés dont il faut s'inspirer pour réaliser la tâche qui nous importe. Enfin peut etre que piloter un moteur pas à pas est suffisamment générique pour que plusieurs personnes l'utilisent. Il ne faudrait pas subdiviser les plugins (et leur noms) jusqu'au point où on aurait "trop" de plugin.

Qu'en dites vous?

**Reply by jerlfan** [reply]
Hello,

I've finally started working on pymodaq_plugins_arduino again.
My philosophy will be as follows. I will consider a "naked" Arduino. 
Inside the plugin I will work on the following examples:
daq_move_stepper_telemetrix
daq_move_output_telemetrix
daq_viewer_input_telemeterix
etc...

These are general uses of the Arduino. Then, for a given "more specific use", you could easily add a plugin based on these, or using pyfirmata, or your own Arduino code: daq_move_USE_ARDUINO-CodeXXX. You could add a thermometer, for example, or a TTL trig, but it will use the output and input anyway, so your plugins could inherit from these plugins.
What do you think of this naming philosophy?

Jérémie

**Reply by seb5g** [reply]
I'm fine with that, we specify a type of use (input, output, stepper...) and a library to do it (telemetrix, pyfirmata...). It would provide a set of base plugins from which a user could build its own specific one

**Reply by seb5g** [reply]
Actually @jerlfan as the last part of the plugin name will also be the class name and the one printed in the control modules, I xould rather you name them:

* daq_move_StepperTelemetrix
* daq_move_OutputTelemetrix
* daq_viewer_InputTelemetrix
